# Module 3: Exercise A

Always load the packages first:

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Read Data to Pandas DataFrame

Let's start by reading the full data set to a pandas DataFrame.

>__Task 1__
>
>Import and explore the data file "diabetes.csv" (Make sure the file is in the same folder with this Jupyter notebook)

In [2]:
diabetes_df = ...

Take a look at the first 10 rows to make an intial check of the data import:

In [ ]:
...

Also, `df.shape` tells you how many rows and columns this data set has:

In [4]:
diabetes_df.shape

(768, 9)

The data has 768 rows and 9 columns corresponding to 768 observations and 9 variables/features.

A more comprehensive check is the `info()` method. Not only does it show the columns and rows, it also shows the data types and memory usage.

In [5]:
diabetes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


---

## Select Columns or Rows

Let's assume we want to select the __Age__ column. We run:

In [6]:
diabetes_df['Age']

0      50
1      31
2      32
3      21
4      33
       ..
763    63
764    27
765    30
766    47
767    23
Name: Age, Length: 768, dtype: int64

The result is NOT a pd.DataFrame but a pd.Series object which is one column of a pd.DataFrame. In other words, __a pd.DataFrame consists of many pd.Series objects.__

Note that in Python, __indices start from 0 (not 1)__. Therefore, if we want the 5th observation, we can write:

In [7]:
diabetes_df['Age'][4]

33

Alternatively, we can use index based location (`df.iloc`) to select rows or columns. This is called "slicing".

__Age__ is the 8th column, so the index is 7:

In [8]:
diabetes_df.iloc[:,7]

0      50
1      31
2      32
3      21
4      33
       ..
763    63
764    27
765    30
766    47
767    23
Name: Age, Length: 768, dtype: int64

Similarly, to get the 5th observation, we write:

In [9]:
diabetes_df.iloc[4,7]

33

>__Task 2__
>
>What is the BMI value of the 55th row? Use either `df[colname][rowname]` or `df.iloc` methods to find it.

In [ ]:
...

### Select More Than One Column

If we want to select more than one column, we must specify the names as another list. Let's assume we want to select the __Glucose__, __BMI__, and __Age__ columns. We can:

1. Plug in the name of the columns as:

In [11]:
cnames = ['Glucose','BMI', 'Age']
cnames

['Glucose', 'BMI', 'Age']

In [12]:
diabetes_df[cnames]

Glucose   BMI  Age
0        148  33.6   50
1         85  26.6   31
2        183  23.3   32
3         89  28.1   21
4        137  43.1   33
..       ...   ...  ...
763      101  32.9   63
764      122  36.8   27
765      121  26.2   30
766      126  30.1   47
767       93  30.4   23

[768 rows x 3 columns]

2. Or directly define the names without creating a list:

In [13]:
diabetes_df[['Glucose','BMI', 'Age']]

Glucose   BMI  Age
0        148  33.6   50
1         85  26.6   31
2        183  23.3   32
3         89  28.1   21
4        137  43.1   33
..       ...   ...  ...
763      101  32.9   63
764      122  36.8   27
765      121  26.2   30
766      126  30.1   47
767       93  30.4   23

[768 rows x 3 columns]

Alternatively, we can use `iloc`:

In [14]:
diabetes_df.iloc[:,[1,5,7]]

Glucose   BMI  Age
0        148  33.6   50
1         85  26.6   31
2        183  23.3   32
3         89  28.1   21
4        137  43.1   33
..       ...   ...  ...
763      101  32.9   63
764      122  36.8   27
765      121  26.2   30
766      126  30.1   47
767       93  30.4   23

[768 rows x 3 columns]

>__Task 3__
>
>Create a dataset of __Insulin__, __Glucose__, and __Outcome__, using both `df[colname][rowname]` and `df.iloc` methods

In [ ]:
...

In [ ]:
...

---

## Find Missing Values

Some missing values are subtle. In our data file, missing values are coded as 0s. We don't expect there to be a person whose __Glucose__, __BloodPressure__, __SkinThickness__, __Insulin__, and __BMI__ values are zero. We can replace these values with "np.NaN":

In [17]:
diabetes_df[['Glucose','BloodPressure','SkinThickness']] = diabetes_df[['Glucose','BloodPressure','SkinThickness']].replace(0,np.NaN)
diabetes_df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0           72.0           35.0        0  33.6   
1              1     85.0           66.0           29.0        0  26.6   
2              8    183.0           64.0            NaN        0  23.3   
3              1     89.0           66.0           23.0       94  28.1   
4              0    137.0           40.0           35.0      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10    101.0           76.0           48.0      180  32.9   
764            2    122.0           70.0           27.0        0  36.8   
765            5    121.0           72.0           23.0      112  26.2   
766            1    126.0           60.0            NaN        0  30.1   
767            1     93.0           70.0           31.0        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

>__Task 4__
>
>There are two more columns (__Insulin__ and __BMI__) where missing values are coded as 0. Replace them with "np.NaN" as well.

In [ ]:
...
diabetes_df

Now, let's use `isna()` to see whether each cell is a missing value or not (`True` for missing value).

In [19]:
diabetes_df.isna()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin    BMI  \
0          False    False          False          False     True  False   
1          False    False          False          False     True  False   
2          False    False          False           True     True  False   
3          False    False          False          False    False  False   
4          False    False          False          False    False  False   
..           ...      ...            ...            ...      ...    ...   
763        False    False          False          False    False  False   
764        False    False          False          False     True  False   
765        False    False          False          False    False  False   
766        False    False          False           True     True  False   
767        False    False          False          False     True  False   

     DiabetesPedigreeFunction    Age  Outcome  
0                       False  False    False  
1                       False  False    False  
2                       False  False    False  
3                       False  False    False  
4                       False  False    False  
..                        ...    ...      ...  
763                     False  False    False  
764                     False  False    False  
765                     False  False    False  
766                     False  False    False  
767                     False  False    False  

[768 rows x 9 columns]

However, we would like to see for each column whether there are any missing values. Try again with the `sum()` method to sum the Boolean values up by columns.

In [20]:
diabetes_df.isna().sum()

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

The `sum()` method can also sum up along the columns (`axis=1`), which we can use to check the number of missing values in any row.

>__Task 5__
>
>Find the number of missing values for each row

In [ ]:
...

---

## Remove Missing Values

To remove missing values, we can use `pd.DataFrame.dropna()`. See [pd.DataFrame.dropna()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html) for the important parameters.

Check again the number of missing values to decide how to deal with each column/variable:

In [22]:
diabetes_df.isna().sum()

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

__Glucose__ has a relatively small portion of missing values, so let's remove the rows with missing values:

In [23]:
diabetes_df.dropna(subset=['Glucose'], inplace=True)

>__Task 6__
>
>- Remove the rows where all of the fields are NA
>- Show the count of the missing values for each column
>- Override the data

In [ ]:
...

>__Task 7__
>
>- Remove the columns containing any NA values
>- Show the count of the missing values for each column
>- Work on "temp"

In [ ]:
# temp for temporary data
temp = diabetes_df...

In [26]:
temp.shape

(763, 5)

---

## Replace Missing Values with Proper Values

The `Dataframe.fillna()` function fills the missing values with a given method. See [pandas.DataFrame.fillna](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html) for the important parameters.

Let's start with replacing values with their averages. Since each column has a different average, we need to do it separately. For example, if we want to replace NAs in __SkinThickness__ with its average, we need to calculate the average first:

In [27]:
diabetes_df['SkinThickness'].mean()

29.149253731343283

Now, let's replace the NAs with the above average:

In [28]:
temp = diabetes_df
temp['SkinThickness'] = temp['SkinThickness'].fillna(temp['SkinThickness'].mean())
temp

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0           72.0      35.000000      NaN  33.6   
1              1     85.0           66.0      29.000000      NaN  26.6   
2              8    183.0           64.0      29.149254      NaN  23.3   
3              1     89.0           66.0      23.000000     94.0  28.1   
4              0    137.0           40.0      35.000000    168.0  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10    101.0           76.0      48.000000    180.0  32.9   
764            2    122.0           70.0      27.000000      NaN  36.8   
765            5    121.0           72.0      23.000000    112.0  26.2   
766            1    126.0           60.0      29.149254      NaN  30.1   
767            1     93.0           70.0      31.000000      NaN  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[763 rows x 9 columns]

>__Task 8__
>
>Fill the missing values with the feature means for __Glucose__ and __BloodPressure__ columns
>
>Assign the data set to "temp" and work on "temp"

In [ ]:
...
temp

Only for learning purposes, we can replace the missing values in a row with the next available data by using `method='bfill'` as below. Note here `inplace=False` tells `fillna` not to make any actual change to the DataFrame. You can also simply remove this argument since `False` is default.

In [30]:
diabetes_df.fillna(method='bfill', limit=10, inplace=False)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0           72.0      35.000000     94.0  33.6   
1              1     85.0           66.0      29.000000     94.0  26.6   
2              8    183.0           64.0      29.149254     94.0  23.3   
3              1     89.0           66.0      23.000000     94.0  28.1   
4              0    137.0           40.0      35.000000    168.0  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10    101.0           76.0      48.000000    180.0  32.9   
764            2    122.0           70.0      27.000000    112.0  36.8   
765            5    121.0           72.0      23.000000    112.0  26.2   
766            1    126.0           60.0      29.149254      NaN  30.1   
767            1     93.0           70.0      31.000000      NaN  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[763 rows x 9 columns]

---

## Create Duplicated Data

The original data does not have any row duplicating altogether. So, we will first create an artificial data set by repeating the same data twice:

In [31]:
diabetes_duplicated = pd.concat([diabetes_df]*2, ignore_index=True)

print(diabetes_df.shape)
print(diabetes_duplicated.shape)

(763, 9)
(1526, 9)


---

## Find Duplicates

`DataFrame.duplicated()` returns a Boolean Series (`True` for duplicate row). 

In [32]:
diabetes_duplicated.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1521     True
1522     True
1523     True
1524     True
1525     True
Length: 1526, dtype: bool

>__Task 9__
>
>Check if __DiabetesPedigreeFunction__ column is duplicated and count the number of duplicates

In [ ]:
...

In [ ]:
...

---

## Remove Duplicate Rows

The above method just locates the duplicate rows. `DataFrame.drop_duplicates()` returns the DataFrame that removes duplicate rows.

See [pandas.DataFrame.drop_duplicates](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html) for the important parameters.

>__Task 10__
>
>Remove duplicate rows and keep the first duplicate row.

In [ ]:
...

Note that `drop_duplicates()` by default returns a non-duplicate DataFrame. However, the original DataFrame is not changed. If you would like to change the original DataFrame, you can add argument `inplace=True`.

We can also remove the rows where a specific column is duplicate by using `drop_duplicates(['columnName'])`.

>__Task 11__
>
>- Remove the rows if duplicates exist in column __DiabetesPedigreeFunction__
>- Keep the last duplicate
>- After you succeed with the above two steps, make the changes directly to the DataFrame

In [ ]:
...

In [ ]:
...